File: internal/server/player_registry.go
package server

import "github.com/lemunozm/ascii-arena/internal/pkg/comm"

type RegistryStatus int

const (
	REGISTRY_OK = RegistryStatus(iota)
	REGISTRY_FULL
	REGISTRY_ALREADY_EXISTS
)

type PlayerRegistry struct {
	maxPlayers  int
	pointsToWin int
	players     []*Player
}

func NewPlayerRegistry(maxPlayers int, pointsToWin int) *PlayerRegistry {
	return &PlayerRegistry{
		maxPlayers:  maxPlayers,
		pointsToWin: pointsToWin,
		players:     []*Player{},
	}
}

func (r *PlayerRegistry) Add(character byte, connection *comm.Connection) RegistryStatus {
	if len(r.players) == r.maxPlayers {
		return REGISTRY_FULL
	}

	for _, p := range r.players {
		if p.GetCharacter() == character {
			return REGISTRY_ALREADY_EXISTS
		}
	}

	player := NewPlayer(character, connection)
	r.players = append(r.players, player)

	return REGISTRY_OK
}

func (r *PlayerRegistry) Remove(player *Player) error {
	//TODO
	return nil
}

func (r PlayerRegistry) GetMaxPlayers() int {
	return r.maxPlayers
}

func (r PlayerRegistry) GetCurrentPlayers() int {
	return len(r.players)
}

func (r PlayerRegistry) GetCharacters() []byte {
	characters := make([]byte, r.GetCurrentPlayers())
	for i, p := range r.players {
		characters[i] = p.GetCharacter()
	}
	return characters
}

func (r PlayerRegistry) IsFull() bool {
	return r.GetMaxPlayers() == r.GetCurrentPlayers()
}

func (r PlayerRegistry) GetPlayers() []*Player {
	return r.players
}

func (r PlayerRegistry) HasWinner() bool {
	for _, p := range r.players {
		if p.GetPoints() >= r.pointsToWin {
			return true
		}
	}
	return false
}

File: src/com/asciiarena/lib/server/player/PlayerRegistry.java
package com.asciiarena.lib.server.player;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import com.asciiarena.lib.common.communication.Connection;
import com.asciiarena.lib.common.communication.ConnectionError;

import javafx.util.Pair;

public class PlayerRegistry
{
    public static enum Status
    {
        OK, 
        FULL, 
        ALREADY_EXISTS;
    }

    private ArrayList<Player> players;
    private final int maxPlayers;
    private final int pointsToWin;

    public PlayerRegistry(int maxPlayers, int pointsToWin)
    {
        this.players = new ArrayList<Player>(maxPlayers);
        this.maxPlayers = maxPlayers;
        this.pointsToWin = pointsToWin;
    }

    public Status add(char character, Connection connection)
    {
        if(players.size() == maxPlayers)
        {
            return Status.FULL;
        }

        for(Player player: players)
        {
            if(player.getCharacter()== character)
            {
                return Status.ALREADY_EXISTS;
            }
        }

        players.add(new Player(character, connection));

        return Status.OK;
    }        

    public void sendToPlayers(Object message) 
    {
        for(Player player: players)
        {
            try
            {
                if(player.isConnected())
                {
                    player.getConnection().send(message);
                }
            }
            catch (ConnectionError e)
            {
                player.markAsDisconnected();
            }
        }
    }

    public List<Pair<Player, Object>> receiveFromPlayers() 
    {
        ArrayList<Pair<Player, Object>> messages = new ArrayList<Pair<Player, Object>>(players.size());
        for(Player player: players)
        {
            try
            {
                Object message = null;
                if(player.isConnected())
                {
                    message = player.getConnection().receive(); 
                }
                messages.add(new Pair<Player, Object>(player, message));
            }
            catch (ConnectionError e)
            {
                player.markAsDisconnected();
            }
        }
        return messages; 
    }

    public boolean hasWinner()
    {
        for(Player player: players)
        {
            if(player.getPoints() >= pointsToWin)
            {
                return true;
            }
        }

        return false;
    }

    public boolean isComplete()
    {
        return players.size() == maxPlayers;
    }

    public List<Character> getCharacters()
    {
        ArrayList<Character> characters = new ArrayList<Character>(players.size());

        for(Player player: players)
        {
            characters.add(player.getCharacter());
        }

        return characters;
    }

    public List<Player> getPlayers()
    {
        return Collections.unmodifiableList(players);
    }

    public int getCurrentPlayers()
    {
        return players.size();
    }

    public int getMaxPlayers()
    {
        return maxPlayers;
    }

    public int getPointsToWin()
    {
        return pointsToWin;
    }
}
